File: copilot-connector/sample_codes/typescript/common-patterns/throttle-resilient-ingestion.ts
// Throttle-resilient ingestion with batch concurrency control
// Shows: retry with exponential backoff on 429, semaphore-based concurrency
//   limiting, and auto-chunking based on payload size
//
// Prerequisites: npm install @microsoft/microsoft-graph-client @azure/identity

import { Client } from "@microsoft/microsoft-graph-client";

/** An external item to ingest into the Microsoft Graph index. */
interface ExternalItem {
  acl: Array<{ type: string; value: string; accessType: string }>;
  properties: Record<string, unknown>;
  content: {
    value: string;
    type: "text" | "html";
  };
}

/** Progress callback for batch ingestion. */
type ProgressCallback = (completed: number, total: number) => void;

/**
 * Provides throttle-resilient item ingestion with retry logic,
 * concurrency control, and automatic content chunking.
 */
class ResilientIngestion {
  constructor(
    private readonly graphClient: Client,
    private readonly connectionId: string
  ) {}

  /**
   * Ingest a single item with exponential backoff retry on throttling.
   *
   * @param itemId - Unique item identifier (must be URL-safe)
   * @param item - The external item payload
   * @param maxRetries - Maximum retry attempts (default: 5)
   */
  async ingestWithRetry(
    itemId: string,
    item: ExternalItem,
    maxRetries = 5
  ): Promise<void> {
    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      try {
        await this.graphClient
          .api(`/external/connections/${this.connectionId}/items/${itemId}`)
          .put(item);
        return; // Success
      } catch (error: unknown) {
        const statusCode = (error as { statusCode?: number }).statusCode;
        if (statusCode !== 429 || attempt === maxRetries) {
          throw error;
        }

        // Prefer Retry-After header; fall back to exponential backoff
        const retryAfterHeader = (error as { headers?: Record<string, string> })
          .headers?.["retry-after"];
        const retryAfterSec = retryAfterHeader
          ? parseInt(retryAfterHeader, 10)
          : Math.pow(2, attempt);

        console.log(
          `Throttled on item ${itemId}. ` +
            `Retry ${attempt + 1}/${maxRetries} after ${retryAfterSec}s`
        );

        await new Promise((resolve) =>
          setTimeout(resolve, retryAfterSec * 1000)
        );
      }
    }
  }

  /**
   * Batch ingest items with controlled concurrency.
   * Uses a semaphore to limit parallel requests (recommended: 4–8,
   * never exceeding the 25 concurrent operations per connection limit).
   *
   * @param items - Map of item ID → external item payload
   * @param maxConcurrency - Maximum parallel requests (default: 4)
   * @param onProgress - Optional progress callback
   */
  async batchIngest(
    items: Map<string, ExternalItem>,
    maxConcurrency = 4,
    onProgress?: ProgressCallback
  ): Promise<void> {
    const semaphore = new Semaphore(maxConcurrency);
    let completed = 0;
    const total = items.size;

    const tasks = Array.from(items.entries()).map(
      async ([itemId, item]) => {
        await semaphore.acquire();
        try {
          await this.ingestWithRetry(itemId, item);
          completed++;
          onProgress?.(completed, total);
        } finally {
          semaphore.release();
        }
      }
    );

    await Promise.all(tasks);
  }

  /**
   * Check payload size before ingestion and auto-chunk if needed.
   * The 4 MB limit applies to the full serialized request body.
   *
   * @param itemId - Base item identifier
   * @param item - The external item payload (may be oversized)
   */
  async ingestWithAutoChunk(
    itemId: string,
    item: ExternalItem
  ): Promise<void> {
    const json = JSON.stringify(item);
    const byteSize = Buffer.byteLength(json, "utf-8");

    if (byteSize > 3_800_000) {
      // Leave 200 KB buffer below the 4 MB limit
      console.log(
        `Item ${itemId} is ${(byteSize / 1_000_000).toFixed(1)}MB — chunking required.`
      );

      const chunks = chunkContent(item.content.value, 3_500_000);

      for (let i = 0; i < chunks.length; i++) {
        const chunkedItem = cloneItemWithChunk(
          itemId,
          item,
          chunks[i],
          i,
          chunks.length
        );
        await this.ingestWithRetry(`${itemId}_chunk_${i}`, chunkedItem);
      }
    } else {
      await this.ingestWithRetry(itemId, item);
    }
  }
}

/**
 * Simple counting semaphore for concurrency control.
 * Limits the number of concurrent async operations.
 */
class Semaphore {
  private current = 0;
  private readonly waiting: Array<() => void> = [];

  constructor(private readonly max: number) {}

  /** Wait until a slot is available. */
  acquire(): Promise<void> {
    if (this.current < this.max) {
      this.current++;
      return Promise.resolve();
    }
    return new Promise<void>((resolve) => {
      this.waiting.push(resolve);
    });
  }

  /** Release a slot, waking the next waiter if any. */
  release(): void {
    const next = this.waiting.shift();
    if (next) {
      next();
    } else {
      this.current--;
    }
  }
}

/**
 * Split content at paragraph boundaries, respecting byte size limits.
 */
function chunkContent(content: string, maxChunkBytes: number): string[] {
  const paragraphs = content.split("\n\n").filter((p) => p.length > 0);
  const chunks: string[] = [];
  let current = "";
  let currentSize = 0;

  for (const para of paragraphs) {
    const paraSize = Buffer.byteLength(para, "utf-8");

    if (currentSize + paraSize > maxChunkBytes && current.length > 0) {
      chunks.push(current);
      current = "";
      currentSize = 0;
    }

    current += para + "\n\n";
    currentSize += paraSize;
  }

  if (current.length > 0) {
    chunks.push(current);
  }

  return chunks;
}

/**
 * Create a chunked version of an item with a unique ID and title context.
 */
function cloneItemWithChunk(
  originalId: string,
  original: ExternalItem,
  chunkContent: string,
  chunkIndex: number,
  totalChunks: number
): ExternalItem {
  const originalTitle =
    (original.properties.title as string) ?? "Document";

  const contextHeader =
    `Document: ${originalTitle}\n` +
    `Part ${chunkIndex + 1} of ${totalChunks}\n\n`;

  return {
    acl: original.acl,
    properties: {
      ...original.properties,
      title: `${originalTitle} (Part ${chunkIndex + 1} of ${totalChunks})`,
      parentDocumentId: originalId,
      chunkIndex,
      totalChunks,
    },
    content: {
      value: contextHeader + chunkContent,
      type: original.content.type,
    },
  };
}

export { ResilientIngestion, Semaphore, chunkContent, ExternalItem };

File: copilot-connector/sample_codes/csharp/common-patterns/throttle-resilient-ingestion.cs
// Throttle-resilient ingestion with batch concurrency control
using System.Text;
using System.Text.Json;
using Microsoft.Graph;
using Microsoft.Graph.Models.ExternalConnectors;

public class ResilientIngestion
{
    private readonly GraphServiceClient _graphClient;
    private readonly string _connectionId;

    public ResilientIngestion(GraphServiceClient graphClient, string connectionId)
    {
        _graphClient = graphClient;
        _connectionId = connectionId;
    }

    /// <summary>
    /// Ingest a single item with exponential backoff retry on throttling.
    /// </summary>
    public async Task IngestWithRetryAsync(ExternalItem item, int maxRetries = 5)
    {
        for (int attempt = 0; attempt <= maxRetries; attempt++)
        {
            try
            {
                await _graphClient.External.Connections[_connectionId]
                    .Items[item.Id]
                    .PutAsync(item);
                return; // Success
            }
            catch (ServiceException ex) when (ex.ResponseStatusCode == 429)
            {
                if (attempt == maxRetries)
                    throw;

                var retryAfter = ex.ResponseHeaders?
                    .RetryAfter?.Delta ?? TimeSpan.FromSeconds(Math.Pow(2, attempt));

                Console.WriteLine(
                    $"Throttled on item {item.Id}. Retry {attempt + 1}/{maxRetries} " +
                    $"after {retryAfter.TotalSeconds}s");

                await Task.Delay(retryAfter);
            }
        }
    }

    /// <summary>
    /// Batch ingest items with controlled concurrency.
    /// Uses a semaphore to limit parallel requests (recommended: 4-8,
    /// never exceeding the 25 concurrent operations per connection limit).
    /// </summary>
    public async Task BatchIngestAsync(
        IEnumerable<ExternalItem> items,
        int maxConcurrency = 4,
        IProgress<(int completed, int total)>? progress = null)
    {
        var itemList = items.ToList();
        var semaphore = new SemaphoreSlim(maxConcurrency);
        int completed = 0;

        var tasks = itemList.Select(async item =>
        {
            await semaphore.WaitAsync();
            try
            {
                await IngestWithRetryAsync(item);
                var count = Interlocked.Increment(ref completed);
                progress?.Report((count, itemList.Count));
            }
            finally
            {
                semaphore.Release();
            }
        });

        await Task.WhenAll(tasks);
    }

    /// <summary>
    /// Check payload size before ingestion and auto-chunk if needed.
    /// The 4 MB limit applies to the full serialized request body.
    /// </summary>
    public async Task IngestWithAutoChunkAsync(ExternalItem item)
    {
        var json = JsonSerializer.Serialize(item);
        var byteSize = Encoding.UTF8.GetByteCount(json);

        if (byteSize > 3_800_000) // Leave 200KB buffer
        {
            Console.WriteLine(
                $"Item {item.Id} is {byteSize / 1_000_000.0:F1}MB — chunking required.");

            var chunks = ChunkContent(item.Content?.Value ?? "", maxChunkBytes: 3_500_000);

            for (int i = 0; i < chunks.Count; i++)
            {
                var chunkedItem = CloneItemWithChunk(item, chunks[i], i, chunks.Count);
                await IngestWithRetryAsync(chunkedItem);
            }
        }
        else
        {
            await IngestWithRetryAsync(item);
        }
    }

    /// <summary>
    /// Split content at paragraph boundaries, respecting byte size limits.
    /// </summary>
    private static List<string> ChunkContent(string content, int maxChunkBytes)
    {
        var paragraphs = content.Split("\n\n", StringSplitOptions.RemoveEmptyEntries);
        var chunks = new List<string>();
        var current = new StringBuilder();
        int currentSize = 0;

        foreach (var para in paragraphs)
        {
            var paraSize = Encoding.UTF8.GetByteCount(para);

            if (currentSize + paraSize > maxChunkBytes && current.Length > 0)
            {
                chunks.Add(current.ToString());
                current.Clear();
                currentSize = 0;
            }

            current.AppendLine(para);
            current.AppendLine();
            currentSize += paraSize;
        }

        if (current.Length > 0)
            chunks.Add(current.ToString());

        return chunks;
    }

    /// <summary>
    /// Create a chunked version of an item with a unique ID and title context.
    /// </summary>
    private static ExternalItem CloneItemWithChunk(
        ExternalItem original, string chunkContent, int chunkIndex, int totalChunks)
    {
        var originalTitle = original.Properties?.AdditionalData?
            .GetValueOrDefault("title")?.ToString() ?? "Document";

        var props = new Dictionary<string, object>(
            original.Properties?.AdditionalData ?? new Dictionary<string, object>());

        props["title"] = $"{originalTitle} (Part {chunkIndex + 1} of {totalChunks})";
        props["parentDocumentId"] = original.Id ?? "";
        props["chunkIndex"] = chunkIndex;
        props["totalChunks"] = totalChunks;

        // Prepend contextual header for self-contained chunks
        var contextHeader =
            $"Document: {originalTitle}\n" +
            $"Part {chunkIndex + 1} of {totalChunks}\n\n";

        return new ExternalItem
        {
            Id = $"{original.Id}_chunk_{chunkIndex}",
            Acl = original.Acl,
            Properties = new Properties { AdditionalData = props },
            Content = new ExternalItemContent
            {
                Value = contextHeader + chunkContent,
                Type = original.Content?.Type ?? ExternalItemContentType.Text
            }
        };
    }
}
